File: src/mia/agents/sender.py
"""
SenderAgent — 消息发送 Agent

职责:
  1. 接收 Scheduler 的 SEND_TEXT / SEND_VOICE 指令
  2. 生成最终回复文本 (SEND_TEXT)
  3. 可选调用 MiMo TTS 生成语音 (SEND_VOICE)
  4. 输出到 CLI 终端
"""

from pathlib import Path
from typing import Optional

from loguru import logger

from mia.agents.base import BaseAgent
from mia.bus.bus import MessageBus
from mia.bus.message import Message, MessageType
from mia.providers.mimo import MiMoProvider


class SenderAgent(BaseAgent):
    """消息发送 Agent — 生成最终回复并输出到用户界面"""

    def __init__(
        self,
        bus: MessageBus,
        mimo: Optional[MiMoProvider] = None,
        output_dir: str = "workspace",
    ):
        """
        Args:
            bus: 消息总线
            mimo: MiMo Provider (用于 TTS 语音合成, 可选)
            output_dir: 语音文件输出目录
        """
        super().__init__(name="sender", bus=bus)
        self.mimo = mimo
        self.output_dir = Path(output_dir)
        self.output_dir.mkdir(parents=True, exist_ok=True)

    async def handle(self, msg: Message) -> None:
        """处理 Scheduler 的发送指令"""
        if msg.msg_type == MessageType.SEND_TEXT:
            await self._handle_send_text(msg)
        elif msg.msg_type == MessageType.SEND_VOICE:
            await self._handle_send_voice(msg)
        elif msg.msg_type == MessageType.STREAM_START:
            await self._handle_stream_start(msg)
        elif msg.msg_type == MessageType.STREAM_CHUNK:
            await self._handle_stream_chunk(msg)
        elif msg.msg_type == MessageType.STREAM_END:
            await self._handle_stream_end(msg)
        else:
            logger.debug("[Sender] 忽略消息类型: {}", msg.msg_type)

    async def _handle_send_text(self, msg: Message) -> None:
        """处理文本发送指令"""
        message = msg.payload.get("message", "")

        # 结构化展示
        from mia.config import get_config
        verbose = get_config().agent.verbose
        if verbose:
            print()
            print(f"\033[32m[Sender]\033[0m 输出回复:")
            print(f"   \033[90m└─\033[0m {message}")
            print()
            print(f"\033[1m{'-'*50}\033[0m")

        logger.info("[Sender] 文本回复已输出, len={}", len(message))

        # 通知 main 对话已完成
        await self.bus.publish(Message(
            msg_type=MessageType.CONVERSATION_DONE,
            source=self.name,
            target="main",
            payload={"message": message},
            session_id=msg.session_id,
        ))

        # 同时通知 MemoryAgent 存储本轮对话
        await self.bus.publish(Message(
            msg_type=MessageType.CONVERSATION_DONE,
            source=self.name,
            target="memory_agent",
            payload={"message": message},
            session_id=msg.session_id,
        ))

    # ─── 流式输出处理 ──────────────────────────────────

    async def _handle_stream_start(self, msg: Message) -> None:
        """流式输出开始 — 打印 header，准备接收增量文本"""
        from mia.config import get_config
        verbose = get_config().agent.verbose
        if verbose:
            print()
            print(f"\033[32m[Sender]\033[0m 输出回复:")
            print(f"   \033[90m└─\033[0m ", end="", flush=True)

    async def _handle_stream_chunk(self, msg: Message) -> None:
        """流式输出文本块 — 立即打印增量文本，不换行"""
        delta = msg.payload.get("delta", "")
        if delta:
            print(delta, end="", flush=True)

    async def _handle_stream_end(self, msg: Message) -> None:
        """流式输出结束 — 打印 footer，发送 CONVERSATION_DONE"""
        message = msg.payload.get("message", "")
        print()  # 流式文本结束，换行
        from mia.config import get_config
        if get_config().agent.verbose:
            print()
            print(f"\033[1m{'-'*50}\033[0m")

        logger.info("[Sender] 流式回复完成, len={}", len(message))

        # 通知 main 对话已完成
        await self.bus.publish(Message(
            msg_type=MessageType.CONVERSATION_DONE,
            source=self.name,
            target="main",
            payload={"message": message},
            session_id=msg.session_id,
        ))

        # 同时通知 MemoryAgent 存储本轮对话
        await self.bus.publish(Message(
            msg_type=MessageType.CONVERSATION_DONE,
            source=self.name,
            target="memory_agent",
            payload={"message": message},
            session_id=msg.session_id,
        ))

    async def _handle_send_voice(self, msg: Message) -> None:
        """处理语音发送指令"""
        message = msg.payload.get("message", "")
        voice = msg.payload.get("voice", "冰糖")
        audio_format = msg.payload.get("format", "wav")

        if not self.mimo:
            logger.warning("[Sender] MiMo Provider 未配置，降级为文本输出")
            await self._handle_send_text(msg)
            return

        # 结构化展示
        print()
        print(f"\033[32m[Sender]\033[0m 输出语音回复 (音色: {voice}):")
        print(f"   \033[90m├─\033[0m 文本: {message}")

        try:
            audio_bytes = await self.mimo.synthesize(
                text=message,
                voice=voice,
                audio_format=audio_format,
            )

            # 保存语音文件
            filename = f"reply_{msg.msg_id}.{audio_format}"
            filepath = self.output_dir / filename
            filepath.write_bytes(audio_bytes)

            print(f"   \033[90m└─\033[0m 语音文件: {filepath}")

            # 自动播放语音 (后台线程，不阻塞)
            import asyncio as _asyncio
            _asyncio.get_event_loop().run_in_executor(
                None, _play_audio_file, str(filepath),
            )
            print()
            print(f"\033[1m{'-'*50}\033[0m")

            logger.info("[Sender] 语音回复已保存: {}", filepath)

        except Exception as e:
            logger.error("[Sender] TTS 合成失败: {}", e)
            print(f"   \033[90m└─\033[0m \033[31m语音合成失败: {e}\033[0m")
            print(f"   \033[90m└─\033[0m 降级为文本: {message}")
            print()
            print(f"\033[1m{'-'*50}\033[0m")

        # 通知 main 对话已完成
        await self.bus.publish(Message(
            msg_type=MessageType.CONVERSATION_DONE,
            source=self.name,
            target="main",
            payload={"message": message},
            session_id=msg.session_id,
        ))

        # 同时通知 MemoryAgent 存储本轮对话
        await self.bus.publish(Message(
            msg_type=MessageType.CONVERSATION_DONE,
            source=self.name,
            target="memory_agent",
            payload={"message": message},
            session_id=msg.session_id,
        ))


# ─── 播放辅助函数 (模块级，供 executor 线程调用) ───

def _play_audio_file(filepath: str) -> None:
    """在后台线程中播放音频文件 — 静默失败不影响主流程"""
    try:
        from mia.audio.playback import play_audio
        play_audio(filepath, blocking=False)
    except Exception:
        pass  # 播放失败不打扰用户

File: src/agents/sender.ts
/**
 * SenderAgent — 消息发送 Agent
 *
 * 职责:
 *   1. 接收 Scheduler 的 SEND_TEXT / SEND_VOICE 指令
 *   2. 生成最终回复文本 (SEND_TEXT)
 *   3. 可选调用 MiMo TTS 生成语音 (SEND_VOICE)
 *   4. 输出到 CLI 终端
 *
 * 与 Python 版 agents/sender.py 保持 1:1 语义映射。
 */

import fs from 'node:fs';
import path from 'node:path';
import { BaseAgent } from './base.js';
import { MessageBus } from '../bus/bus.js';
import { Message, MessageType } from '../bus/message.js';
import { MiMoProvider } from '../providers/mimo.js';
import { getConfig } from '../config.js';

/**
 * SenderAgent — 生成最终回复并输出到用户界面
 */
export class SenderAgent extends BaseAgent {
  private mimo: MiMoProvider | null;
  private outputDir: string;

  constructor(
    bus: MessageBus,
    mimo: MiMoProvider | null = null,
    outputDir = 'workspace',
  ) {
    super('sender', bus);
    this.mimo = mimo;
    this.outputDir = path.resolve(outputDir);
    fs.mkdirSync(this.outputDir, { recursive: true });
  }

  /** 消息分发 */
  protected async handle(msg: Message): Promise<void> {
    switch (msg.msg_type) {
      case MessageType.SEND_TEXT:
        await this._handleSendText(msg);
        break;
      case MessageType.SEND_VOICE:
        await this._handleSendVoice(msg);
        break;
      case MessageType.STREAM_START:
        await this._handleStreamStart(msg);
        break;
      case MessageType.STREAM_CHUNK:
        await this._handleStreamChunk(msg);
        break;
      case MessageType.STREAM_END:
        await this._handleStreamEnd(msg);
        break;
    }
  }

  /** 处理文本发送指令 */
  private async _handleSendText(msg: Message): Promise<void> {
    const message = (msg.payload['message'] as string) || '';

    // 结构化展示
    const verbose = getConfig().agent.verbose;
    if (verbose) {
      console.log();
      console.log(`\x1b[32m[Sender]\x1b[0m 输出回复:`);
      console.log(`   \x1b[90m└─\x1b[0m ${message}`);
      console.log();
      console.log(`\x1b[1m${'-'.repeat(50)}\x1b[0m`);
    }

    // 通知 main 对话已完成 + MemoryAgent 存储
    await this._emitConversationDone(message, msg.session_id);
  }

  // ─── 流式输出处理 ──────────────────────────────

  /** 流式输出开始 */
  private async _handleStreamStart(_msg: Message): Promise<void> {
    const verbose = getConfig().agent.verbose;
    if (verbose) {
      console.log();
      console.log(`\x1b[32m[Sender]\x1b[0m 输出回复:`);
      process.stdout.write(`   \x1b[90m└─\x1b[0m `);
    }
  }

  /** 流式输出文本块 — 立即打印增量文本 */
  private async _handleStreamChunk(msg: Message): Promise<void> {
    const delta = (msg.payload['delta'] as string) || '';
    if (delta) {
      process.stdout.write(delta);
    }
  }

  /** 流式输出结束 */
  private async _handleStreamEnd(msg: Message): Promise<void> {
    const message = (msg.payload['message'] as string) || '';
    console.log(); // 流式结束换行

    const verbose = getConfig().agent.verbose;
    if (verbose) {
      console.log();
      console.log(`\x1b[1m${'-'.repeat(50)}\x1b[0m`);
    }

    await this._emitConversationDone(message, msg.session_id);
  }

  /** 处理语音发送指令 */
  private async _handleSendVoice(msg: Message): Promise<void> {
    const message = (msg.payload['message'] as string) || '';
    const voice = (msg.payload['voice'] as string) || '冰糖';
    const audioFormat = (msg.payload['format'] as string) || 'wav';

    if (!this.mimo) {
      // 无 TTS Provider，降级为文本
      console.warn('[Sender] MiMo Provider 未配置，降级为文本输出');
      await this._handleSendText(msg);
      return;
    }

    console.log();
    console.log(`\x1b[32m[Sender]\x1b[0m 输出语音回复 (音色: ${voice}):`);
    console.log(`   \x1b[90m├─\x1b[0m 文本: ${message}`);

    try {
      const audioBytes = await this.mimo.synthesize(
        message,
        voice,
        audioFormat,
      );

      // 保存语音文件
      const filename = `reply_${msg.msg_id}.${audioFormat}`;
      const filepath = path.join(this.outputDir, filename);
      fs.writeFileSync(filepath, audioBytes);

      console.log(`   \x1b[90m└─\x1b[0m 语音文件: ${filepath}`);
      console.log();
      console.log(`\x1b[1m${'-'.repeat(50)}\x1b[0m`);
    } catch (err) {
      console.error(`[Sender] TTS 合成失败:`, err);
      console.log(`   \x1b[90m└─\x1b[0m \x1b[31m语音合成失败: ${err}\x1b[0m`);
      console.log(`   \x1b[90m└─\x1b[0m 降级为文本: ${message}`);
      console.log();
      console.log(`\x1b[1m${'-'.repeat(50)}\x1b[0m`);
    }

    await this._emitConversationDone(message, msg.session_id);
  }

  /**
   * 发送 CONVERSATION_DONE 给 main 和 memory_agent
   * 双目标投递确保主循环退出 + 记忆存储
   */
  private async _emitConversationDone(
    message: string,
    sessionId?: string,
  ): Promise<void> {
    const doneMsg: Message = {
      msg_type: MessageType.CONVERSATION_DONE,
      source: this.name,
      target: 'main',
      payload: { message },
      msg_id: Date.now().toString(16),
      timestamp: Date.now(),
      session_id: sessionId,
    };

    // 通知 main
    await this.bus.publish(doneMsg);

    // 通知 MemoryAgent (双目标)
    await this.bus.publish({
      ...doneMsg,
      target: 'memory_agent',
      msg_id: Date.now().toString(16) + 'm',
    });
  }
}
